pallets/assets-transaction-payment/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+			Ok((fee, InitialPayment::Nothing))
+		} else if self.asset_id != Zero::zero() {
+			// If the asset id is zero, then we treat that case as payment in PDEX,
+			T::OnChargeAssetTransaction::withdraw_fee(
+				who,
+				call,
+				info,
+				self.asset_id,
+				fee.into(),
+				self.tip.into(),
+			)
+			.map(|i| (fee, InitialPayment::Asset(i.into())))
+		} else {
+			<OnChargeTransactionOf<T> as OnChargeTransaction<T>>::withdraw_fee(
+				who, call, info, fee, self.tip,
+			)
+			.map(|i| (fee, InitialPayment::Native(i)))
+			.map_err(|_| -> TransactionValidityError { InvalidTransaction::Payment.into() })
+		}
+	}
+}
+
+impl<T: Config> sp_std::fmt::Debug for ChargeAssetTransactionPayment<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(f, "ChargeTransactionPayment<{self:?}>")
+	}
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: Config> SignedExtension for ChargeAssetTransactionPayment<T>
+where
+	T::Call: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+	AssetBalanceOf<T>: Send + Sync + FixedPointOperand,
+	BalanceOf<T>: Send + Sync + From<u64> + FixedPointOperand + IsType<ChargeAssetBalanceOf<T>>,
+	ChargeAssetIdOf<T>: Send + Sync + Zero,
+	CreditOf<T::AccountId, T::Fungibles>: IsType<ChargeAssetLiquidityOf<T>>,
+{
+	const IDENTIFIER: &'static str = "AssetsTransactionPayment";
+	type AccountId = T::AccountId;
+	type Call = T::Call;
+	type AdditionalSigned = u8;
+	type Pre = (
+		// tip
+		BalanceOf<T>,
+		// who paid the fee - this is an option to allow for a Default impl.
+		T::AccountId,
+		// imbalance resulting from withdrawing the fee
+		InitialPayment<T>,
+	);
+
+	fn additional_signed(&self) -> Result<Self::AdditionalSigned, TransactionValidityError> {
+		Ok(self.signature_scheme)
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> TransactionValidity {
+		use pallet_transaction_payment::ChargeTransactionPayment;
+		let (fee, inital_payment) = self.withdraw_fee(who, call, info, len)?;
+		// Check if the given asset is valid
+		if let InitialPayment::Asset(asset) = inital_payment {
+			let allowed_assets = <AllowedAssets<T>>::get();
+			if !allowed_assets.contains(&asset.asset()) {
+				return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+			}
+		}
+		let priority = ChargeTransactionPayment::<T>::get_priority(info, len, self.tip, fee);
+		Ok(ValidTransaction { priority, ..Default::default() })
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		call: &Self::Call,
+		info: &DispatchInfoOf<Self::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
+		// Check if the given asset is valid
+		if let InitialPayment::Asset(asset) = &initial_payment {
```

Comment:
Added a custom error

Reply:
Which caller function? This function should be safe to call without outer scope. Assertion/error should be returned if its not valid since its even not validates asset because of this if...